### PR TITLE
NIFI-4946 nifi-spark-bundle : Adding support for pyfiles, file, jars options

### PR DESCRIPTION
**Pull request by Mageswaran1989**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by Mageswaran1989** [comment]
Team,

This MR was created in line with our current requirements. 
Currently the changes are tested manually and found working.

We would like to make this changes go in mainline, for which we need community help with reviewing and adding test code, since we are new to the Nifi extensions.

Next plan is to add test cases : 
- Create a sample PySpark example modules as part of Test resources 
- Use that in tests

Could any one point out some existing test cases that does similar testing.

File: nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java [hunk 1]
```diff
@@ -178,48 +247,188 @@ public void onTrigger(ProcessContext context, final ProcessSession session) thro
...
+                            session.transfer(flowFile, REL_WAIT);
+                            break;
+                    }
+
+                } catch (JSONException | InterruptedException e) {
+
+                    //Incoming flow file is not an JSON file hence consider it to be an triggering point
+
+                    String batchPayload = "{ \"pyFiles\": [\"" +context.getProperty(PY_FILES).getValue()+ "\"], " +
+                            "\"file\" : \""+context.getProperty(MAIN_PY_FILE).getValue()+"\" }";
```
Comment (zenfenan) [inline nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java]: This is confusing to me. Why are we saying that if the incoming flowfile is not a valid JSON, we are going ahead with the assumption that it is going to be PySpark? I mean the assumption here lacks clarity. Please correct me, if I'm wrong.

File: nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java [hunk 2]
```diff
@@ -178,48 +247,188 @@ public void onTrigger(ProcessContext context, final ProcessSession session) thro
...
+                            break;
+                        default:
+                            log.debug(" ====> default State: " + state);
+                            session.transfer(flowFile, REL_WAIT);
+                            break;
+                    }
+
+                } catch (JSONException | InterruptedException e) {
+
+                    //Incoming flow file is not an JSON file hence consider it to be an triggering point
```
Comment (zenfenan) [inline nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java]: Cosmetic change: Multiple empty lines were left. IMHO, one empty line should be enough for better readability.

File: nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java [hunk 3]
```diff
@@ -129,7 +185,13 @@
     public void init(final ProcessorInitializationContext context) {
         List<PropertyDescriptor> properties = new ArrayList<>();
         properties.add(LIVY_CONTROLLER_SERVICE);
+        properties.add(IS_BATCH_JOB);
+        properties.add(PY_FILES);
+//        properties.add(JAR_FILES);
+        properties.add(MAIN_PY_FILE);
+        properties.add(NAME);
         properties.add(CODE);
+//        properties.add(ARGS);
```
[the remark this answers is not in the document]
Comment (Mageswaran1989) [reply nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java]: Only `pyfiles` and `file` options are tested. Rest are yet to be tested.

Plan was to go with implementing test modules and test other features, since the current manual testing takes a long routine of compile, copy and restart of the Nifi.

File: nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java [hunk 4]
```diff
@@ -178,48 +247,188 @@ public void onTrigger(ProcessContext context, final ProcessSession session) thro
...
+                        jsonResponse = IOUtils.toString(inputStream, charset);
+                    } catch (IOException ioe) {
+                        log.error("Error reading input flowfile, penalizing and routing to failure", new Object[]{flowFile, ioe.getMessage()}, ioe);
+                        flowFile = session.penalize(flowFile);
+                        session.transfer(flowFile, REL_FAILURE);
+                        return;
+                    }
+                }
+
+                log.debug(" ====> jsonResponse: " + jsonResponse);
```
[the remark this answers is not in the document]
Comment (Mageswaran1989) [reply nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java]: Sure, this will be removed in the next commit.

File: nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java [hunk 5]
```diff
@@ -83,6 +83,62 @@
...
+            .displayName("pyFiles")
+            .description("Python files to be used in this batch session that includes *.py, *.zip files")
+            .required(false)
+            .addValidator(StandardValidators.createURLorFileValidator())
+            .expressionLanguageSupported(false)
+            .build();
+
+    public static final PropertyDescriptor JAR_FILES =  new PropertyDescriptor.Builder()
+            .name("exec-spark-iactive-jarfiles")
+            .displayName("jars")
```
[the remark this answers is not in the document]
Comment (Mageswaran1989) [reply nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java]: Sure, will do that.

File: nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java [hunk 6]
```diff
@@ -83,6 +83,62 @@
...
+            .description("jars to be used in this batch session")
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .expressionLanguageSupported(false)
+            .build();
+
+    public static final PropertyDescriptor NAME =  new PropertyDescriptor.Builder()
```
[the remark this answers is not in the document]
Comment (Mageswaran1989) [reply nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java]: Like said before, not yet considered. We just wanted to get a hang of the code with our basic requirements

File: nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java [hunk 7]
```diff
@@ -178,48 +247,188 @@ public void onTrigger(ProcessContext context, final ProcessSession session) thro
...
-            log.debug("ExecuteSparkInteractive Result of Job Submit: " + result);
-            if (result == null) {
-                session.transfer(flowFile, REL_FAILURE);
-            } else {
+
+            if (isBatchJob) {
+
+                String jsonResponse = null;
+
+                if (StringUtils.isEmpty(jsonResponse)) {
```
[the remark this answers is not in the document]
Comment (Mageswaran1989) [reply nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java]: Once current approach is accepted this can be taken care

File: nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java (hunk 1 above)
Comment (Mageswaran1989) [reply nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java]: Could you please check the description @ https://issues.apache.org/jira/browse/NIFI-4946

The assumption was made such that it doesn't break existing code flow and at the same time we wanted to know the status of the submitted job.

So the naive idea was to re-route the Livy Json response back to Spark processor only, so that it can get last submitted `url` from the custom (tampered) JSON response, wait for user specified wait time and again query the Livy for the Job status in a loop till it succeeds or fails.

So when the processor is configured to submit a Spark job, it will expect the incoming flowfile to be an custom Json response with an `url` field to query the Livy, if not it is considered as a triggering point nothing else.

I am open for any ideas from your end.

Thanks.

Comment (Mageswaran1989) [reply nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java]: @zenfenan could please review above logic and suggest a way to handle plain Scala/Python code and packages source files for pyfiles/jars ?

Comment (zenfenan) [reply nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java]: The code can be submitted using the `CODE` property, right? That used to work. Or are you asking of a way to upload/send files or jars through Livy?

Comment (Mageswaran1989) [reply nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java]: As per the code flow @ https://issues.apache.org/jira/browse/NIFI-4946, currently I am able to send *.zip files (Python modules) through livy. My question was what should we do with flowfile, when we are using the processor to submit a batch job?

Comment (Mageswaran1989) [reply nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java]: Sometime this week I am planning to add support for jar files, args and application name over the Livy options.

The catch here is unlike plain Spark code, batch process code will take time to finish which is expected one as we know. So as a hack I was re-routing the Json response after batch submission to itself, where I poll the incoming flowfile and check whether it is a Json file and if so I will try to get the "livy url" to post again to know the status of the batch job as long as it runs. After knowing the the job finished, the success route is triggered.

That was the reason the I have made an assumption if the incoming file is Json, it is from previous batch job submission.

In short, the flow file :
- Is considered as a triggering point (or)
- Is considered as plain Spark code that compiles over Livy (or)
- Is a Livy Json response, which can further be used to check the status of long running Spark batch job

I was looking for the right Nifi way of handling this? 

I feel I am too conservative and trying to fit all the functionalities on one processor.
- Flow file/property can be used to run a Spark code
- Pyfiles can be used to run Spark batch job
- Jars can be used to run Spark batch job
- Args options for batch mode
- By rerouting the success to itself, it can monitor the long running job over Livy rest APIs

Comment (mattyb149) [reply nifi-nar-bundles/nifi-spark-bundle/nifi-livy-processors/src/main/java/org/apache/nifi/processors/livy/ExecuteSparkInteractive.java]: Since the current processor is called ExecuteSparkInteractive, you could move the batch functionality out into something called SubmitSparkJob or something. The outgoing flow file could contain the application ID and/or any other information that would allow you to monitor the job downstream (perhaps with InvokeHttp through Livy, e.g.) Are you still interested in working on this? It seems like very nice functionality to have in NiFi!
